File: packages/moss-repo-indexer/src/moss_repo_indexer/sync.py
"""End-to-end: resolve source → build documents → upload (unless dry_run)."""

from __future__ import annotations

from dataclasses import dataclass, replace
from typing import List, Optional

from moss import DocumentInfo, MutationResult

from .clone import ResolvedSource, resolve_source
from .documents import build_documents
from .types import IndexOptions, SyncOptions
from .uploader import UploadOptions, upload_documents


@dataclass
class SyncResult:
    documents: List[DocumentInfo]
    mutation: Optional[MutationResult]
    dry_run: bool
    repo_name: str
    root: str


async def sync(options: SyncOptions) -> SyncResult:
    """Index a local path or git URL into Moss."""
    resolved = resolve_source(options.source, options.index.ref)
    try:
        return await _sync_resolved(options, resolved)
    finally:
        resolved.close()


async def _sync_resolved(options: SyncOptions, resolved: ResolvedSource) -> SyncResult:
    index_opts = _with_repo_name(options.index, resolved.repo_name)
    documents = build_documents(resolved.root, index_opts)
    if options.dry_run:
        return SyncResult(
            documents=documents,
            mutation=None,
            dry_run=True,
            repo_name=resolved.repo_name,
            root=str(resolved.root),
        )

    mutation = await upload_documents(
        documents,
        options.creds,
        UploadOptions(upsert=options.upsert),
    )
    return SyncResult(
        documents=documents,
        mutation=mutation,
        dry_run=False,
        repo_name=resolved.repo_name,
        root=str(resolved.root),
    )


def _with_repo_name(options: IndexOptions, repo_name: str) -> IndexOptions:
    if options.repo_name:
        return options
    return replace(options, repo_name=repo_name)

File: packages/moss-repo-indexer-js/src/sync.ts
import type { RepoDocument } from './document.js'
import { resolveSource, type ResolvedSource } from './clone.js'
import { buildDocuments } from './documents.js'
import { SyncOptions, resolveIndexOptions, type IndexOptions } from './types.js'
import { uploadDocuments, type MutationResultLike } from './uploader.js'

export interface SyncResult {
  documents: RepoDocument[]
  mutation?: MutationResultLike
  dryRun: boolean
  repoName: string
  root: string
}

export async function sync(options: SyncOptions): Promise<SyncResult> {
  const index = resolveIndexOptions(options.index ?? {})
  const resolved = resolveSource(options.source, index.ref)
  try {
    return await syncResolved(options, resolved, index)
  } finally {
    resolved.close()
  }
}

async function syncResolved(
  options: SyncOptions,
  resolved: ResolvedSource,
  index: ReturnType<typeof resolveIndexOptions>,
): Promise<SyncResult> {
  const indexOpts: IndexOptions = {
    ...index,
    repoName: index.repoName ?? resolved.repoName,
  }
  const documents = buildDocuments(resolved.root, indexOpts)
  if (options.dryRun) {
    return {
      documents,
      dryRun: true,
      repoName: resolved.repoName,
      root: resolved.root,
    }
  }
  const mutation = await uploadDocuments(documents, options.creds, {
    upsert: options.upsert ?? false,
    replace: options.replace ?? false,
  })
  return {
    documents,
    mutation,
    dryRun: false,
    repoName: resolved.repoName,
    root: resolved.root,
  }
}
